### ISSUE TITLE
Fleet-Controller is consuming high CPU

### ISSUE THREAD

**Issue by hv3374** [rancher/fleet#2108]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

We have a fleet controller managing multiple downstream clusters to push new releases in the form of bundles.
It is deployed in the **local** cluster and consumes a high amount of CPU, further impacting the host node and making it inaccessible.

<img width="692" alt="image" src="https://github.com/rancher/fleet/assets/53660076/897420f0-5d9e-4db5-a82c-80c40fb2f785">

These are the messages we are seeing in the fleet logs:

```
time="2024-01-30T15:56:46Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.099183146s"
time="2024-01-30T15:56:46Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:eebe75a264385f7d80cf5fed78d2aea0d1aeb420]) 1.094558203s"
time="2024-01-30T15:56:46Z" level=info msg="rate limited bundle(map[objectset.rio.cattle.io/hash:f11f48ca74e08814cf9e43c95bfef0a81e4223b1]) 1.001122105s"
time="2024-01-30T15:56:48Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.007134158s"
time="2024-01-30T15:56:54Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:eebe75a264385f7d80cf5fed78d2aea0d1aeb420]) 1.026795575s"
time="2024-01-30T15:56:59Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.017403089s"
time="2024-01-30T15:57:01Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.018267701s"
time="2024-01-30T15:57:03Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.213785793s"
time="2024-01-30T15:57:03Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.052259493s"
time="2024-01-30T15:57:12Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.002590941s"
time="2024-01-30T15:57:15Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.009021079s"
time="2024-01-30T15:57:16Z" level=info msg="rate limited bundle(map[objectset.rio.cattle.io/hash:f11f48ca74e08814cf9e43c95bfef0a81e4223b1]) 1.111663992s"
time="2024-01-30T15:57:16Z" level=info msg="rate limited bundle(map[objectset.rio.cattle.io/hash:999a2834038888622538111b467e85a1e2f58489]) 1.034384444s"
time="2024-01-30T15:57:16Z" level=info msg="rate limited bundle(map[objectset.rio.cattle.io/hash:f35215d0b6a63e71cf17d96658090308a1c458b2]) 1.079223512s"
time="2024-01-30T15:57:17Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.930740235s"
time="2024-01-30T15:57:17Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.136811564s"
time="2024-01-30T15:57:21Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.02522063s"
time="2024-01-30T15:57:21Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.073943352s"
time="2024-01-30T15:57:25Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.058486832s"
time="2024-01-30T15:57:28Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:eebe75a264385f7d80cf5fed78d2aea0d1aeb420]) 1.07143577s"
time="2024-01-30T15:57:28Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.813639181s"
time="2024-01-30T15:57:28Z" level=info msg="rate limited bundle(map[objectset.rio.cattle.io/hash:75c68f740e6a4200fc37547a13696fd2b38cb0aa]) 1.584292178s"
time="2024-01-30T15:57:28Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.057524831s"
time="2024-01-30T15:57:30Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.012235596s"
time="2024-01-30T15:57:32Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.003111633s"
time="2024-01-30T15:57:33Z" level=info msg="rate limited bundle(map[objectset.rio.cattle.io/hash:999a2834038888622538111b467e85a1e2f58489]) 1.002578045s"
time="2024-01-30T15:57:39Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.286965846s"
time="2024-01-30T15:57:39Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.284870178s"
time="2024-01-30T15:57:39Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:eebe75a264385f7d80cf5fed78d2aea0d1aeb420]) 1.557274265s"
time="2024-01-30T15:57:48Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.076845886s"
time="2024-01-30T15:57:52Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:eebe75a264385f7d80cf5fed78d2aea0d1aeb420]) 1.019249176s"
time="2024-01-30T15:57:53Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:6e6a85a80e10dd25cd481e95c24f43912815c06f]) 1.133823301s"
time="2024-01-30T15:57:59Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:36f7a31adbf4d654c41b791d018e09b831cb5e65]) 1.063237675s"
time="2024-01-30T15:57:59Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7b4ecfeafb16892266b652087343e35af27691aa]) 1.036100606s"
``` 

Is the high CPU consumption expected? If not, Is there anything we can do from our side to remediate this?
  

### Expected Behavior

The fleet Controller consumption CPU shouldn't be so high.
As per the patches implemented in previous versions, this issue shouldn't occur again: https://github.com/rancher/fleet/issues/721#issuecomment-1522944898

### Steps To Reproduce

Connect fleet to multiple Git Repos associated with multiple downstream clusters

### Environment

```markdown
- Architecture:x86-64
- Fleet Version:0.8.0
- Cluster: local
  - Provider: Rancher v2.7.6
  - Options: 2 AWS ubuntu nodes, m5a.xlarge each
  - Kubernetes Version: v1.26.10+k3s1
```

### Anything else?

This behaviour has been seen in the fleet-controller that has been managing multiple git repos and clusters (more than 10)

**Comment by kkaempf** [comment]
/cc @aruiz14 

@hv3374 can you give [v0.8.2-rc.3](https://github.com/rancher/fleet/releases/tag/v0.8.2-rc.3) a try ?

**Comment by hv3374** [comment]
@

> /cc @aruiz14
> 
> @hv3374 can you give [v0.8.2-rc.3](https://github.com/rancher/fleet/releases/tag/v0.8.2-rc.3) a try ?

Thanks, @kkaempf, 
Tested it in a dev environment, and the CPU consumption by fleet has significantly reduced.

Unfortunately, We can't deploy an RC version in our **PROD** environment.

When can we expect a stable release for production?

**Comment by kkaempf** [comment]
see #1842, #2054

closing as duplicate

**Comment by kkaempf** [comment]
@hv3374 this fix should become available on the next 2.7.x and 2.8.x releases.
